docs/features/role-based-access-control/high-level-design.md

```diff
@@ -0,0 +1,100 @@
+# Role base access control - High-level design
+
+This document aims at providing an overview of the architecture of Role-based Access Control (RBAC from here on) for
+the Marxan platform.
+
+## Roles
+
+Platform administrators should be identified by a boolean attribute in the users table (`is_admin`). This attribute can
+only be edited by other platform administrators, and it must not be editable by the user when editing their profile.
+
+The remaining roles supported by the platform (`Owner`, `Contributor` and `Viewer`) will be stored in the `roles` 
+database in the `marxan-api` database.
+
+Roles should have numerical values associated to them, to identify the hierarchical order between them. For instance:
+
+* `Owner` -> 1000
+* `Contributor` -> 500
+* `Viewer` -> 250
+
+Roles with fewer privileges should have lower numerical values than roles with more privileges. Using numerical values
+for roles allows for easily comparing if a given user has certain privileges (e.g. checking if a user has the role 
+`Contributor` can be easily done: `role >= Contributor (500)`).
```

Comment:
Although I like the idea (deriving from rwx?), it can be misleading as we may have not overlapping capabilities - for example, how would it look when:|
* role A allows to view only solutions
* role B allows to view only scenario setup

Can be compose such values in an obvious and clear way, that "computation" on numerical values would give us relevant access to what user should be allowed, denying the other?

Reply:
I'm not sure I understand your question... Why are the roles you described overlapping? They are different resources (solutions vs scenarios), so they would have to be subject to different permission control, right? You could, for instance, have role "Viewer" for solutions, and role "Owner" for scenarios, or any other combination of roles...

Composition/inheritance of permissions only comes into play when you don't explicitly set anything, so if you want to explicitly allow something and deny others, you can do it in the corresponding DB tables: for instance, you could be "Owner" of a "solution" (for instance) and just "Viewer" of a scenario (though the current document is not contemplating permissions for solutions)

Reply:
@ikas so, solutions are different resource ? It wasn't explicite said anywhere ;) 

we may also try another example:
* role A allows to edit inclusion/exclusion of PU within scenario
* role B allows to edit cost surface within scenario
* role C allows full edit on any scenario aspect
* ... and so on.

I.e. the scope (A & B) isn't overlapping (but related to the same resource - examples are there just to visualize, don't focus on those too much). How would the proposed approach support this?

Reply:
Hum, I assumed scenarios and solutions would be the same thing, I think that's what I got wrong :sweat_smile:

Let me have a think and circle back!

Reply:
looking at this from a different angle, I recommend to not hardcode in the design the assumption that different roles are supersets or subsets of another role.

I think that avoiding this assumption can give more flexibility in the modelling of access control, at the expense of more verbosity in the "roles X permissions" setup

That is, if we cannot assume that "Owner" is a proper superset of "Contributor", then we'd need to assign to "Owner" any permissions that we want the role to have from those of "Contributor", which may even be all of them, but since we'd do this once and we will only have a small handful of roles, I think the benefits of allowing for only partly-intersecting sets outweigh the convenience of defining roles as supersets/subsets.

Reply:
@hotzevzl I agree, I was trying to get to that point in longer path ;) Inheritance overall may lead to more confusion the more complicated privileges we have (keeping in mind that our system is not JIRA and we don't have such team and years for development)

Reply:
As discussed during today's daily meeting, I'll remove the "hierarchical" bits from this implementation.